src/server/coprocessor/endpoint.rs

```diff
@@ -486,14 +480,163 @@ fn get_chunk(chunks: &mut Vec<Chunk>) -> &mut Chunk {
+                        -> Result<()> {
+    for col in cols {
+        let col_id = col.get_column_id();
+        if let Some(v) = values.get(&col_id) {
+            data.extend_from_slice(v);
+            continue;
+        }
+        if col.get_pk_handle() {
+            box_try!(datum::encode_to(data, &[get_pk(col, h)], false));
+        } else if mysql::has_not_null_flag(col.get_flag() as u64) {
+            return Err(box_err!("column {} of {} is missing", col_id, h));
+        } else {
+            box_try!(datum::encode_to(data, &[Datum::Null], false));
+        }
+    }
+    Ok(())
+}
+
+#[derive(Debug, Clone)]
+struct SortRow {
+    key: Vec<Datum>,
+    handle: i64,
+    data: Vec<u8>,
+    order_cols: Vec<ByItem>,
+    ctx: Rc<EvalContext>,
+}
+
+
+impl SortRow {
+    fn new(handle: i64,
+           order_cols: Vec<ByItem>,
+           ctx: Rc<EvalContext>,
+           data: Vec<u8>,
+           key: Vec<Datum>)
+           -> SortRow {
+        SortRow {
+            key: key,
+            handle: handle,
+            data: data,
+            order_cols: order_cols,
+            ctx: ctx,
+        }
+    }
+
+    fn cmp_and_check(&self, right: &SortRow) -> Result<CmpOrdering> {
+        let values = self.key.iter().zip(right.key.iter());
+        for (col, (v1, v2)) in self.order_cols.iter().zip(values) {
+            let cmp_ret = v1.cmp(self.ctx.as_ref(), v2);
+            if let StdError::Err(err) = cmp_ret {
+                return Err(box_err!("cmp failed {:?} ", err));
+            }
+
+            let order = cmp_ret.unwrap();
+            if order == CmpOrdering::Equal {
+                continue;
+            }
+            // less or equal
+            if col.get_desc() {
+                // heap pop the biggest first
+                return Ok(order.reverse());
+            } else {
+                return Ok(order);
+            }
+        }
+        Ok(CmpOrdering::Equal)
+    }
+}
+
+struct TopNHeap {
+    rows: BinaryHeap<SortRow>,
+    limit: usize,
+}
+
+impl TopNHeap {
+    fn new(limit: usize) -> TopNHeap {
+        TopNHeap {
+            rows: match limit {
+                usize::MAX => BinaryHeap::new(),
+                _ => BinaryHeap::with_capacity(limit + 1),
+            },
+            limit: limit,
+        }
+    }
+
+    fn try_add_row(&mut self, row: SortRow) -> Result<()> {
+        if self.rows.is_empty() {
+            self.rows.push(row);
+            return Ok(());
+        }
+
+        // check row data
+        let order = {
+            let top_data = self.rows.peek().unwrap();
+            try!(row.cmp_and_check(&top_data))
+        };
+
+        // push into heap on heap not full
+        if self.rows.len() < self.limit {
+            self.rows.push(row);
+            return Ok(());
+        }
+
+        // switch top value with row when heap is full and current row is smaller than top data
+        let mut top_value = self.rows.peek_mut().unwrap();
```

Comment:
Assume limit is 3, the data here is now 1, 2, 5, if we inert 3 here, seem we should remove 5 and add 3.